### PR TITLE
Stop sharing when Tor stops during startup

### PR DESCRIPTION
**Pull request by grote**
Each manager is now publishing its own state flow and the share manager combines all three into one sharing state flow, so it can update that state even after sharing is complete and the onionshare was published.

Related to #12

### CODE REVIEW

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt [hunk 1]
```diff
@@ -66,10 +66,14 @@ class TorManager @Inject constructor(
                     LOG.debug("TorService: Started")
                     startLatch?.countDown() ?: LOG.error("startLatch was null when Tor started.")
                 }
-                // FIXME When we stop unplanned, we need to inform the ShareManager
-                //  that we stopped, so it can clear its state up, stopping webserver, etc.
-                TorService.STATUS_STOPPING -> LOG.debug("TorService: Stopping...")
-                TorService.STATUS_OFF -> LOG.debug("TorService: Stopped")
+                TorService.STATUS_STOPPING -> {
+                    LOG.debug("TorService: Stopping...")
+                    _state.value = TorState.Stopping
```
Comment (akwizgran) [inline app/src/main/java/org/onionshare/android/tor/TorManager.kt:71]: Looks sensible to reflect the status in our own StateFlow. Should we do the same for STATUS_STARTING and STATUS_ON?

File: app/src/main/java/org/onionshare/android/ShareManager.kt [hunk 2]
```diff
@@ -126,12 +126,15 @@ class ShareManager @Inject constructor(
                         _shareState.value =
                             ShareUiState.Starting(files, 100, state.progress)
                         onion = state.onion
+                    } else if (state is TorState.Stopping || state is TorState.Stopped) {
+                        // FIXME When Tor stops unplanned *after* starting,
+                        //  we also need to clear up our state, stopping webserver, etc.
+                        throw IOException("Tor has stopped unexpectedly")
```
Comment (akwizgran) [inline app/src/main/java/org/onionshare/android/ShareManager.kt:132]: I was able to trigger this by repeatedly tapping the start/stop button - is that expected?

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt (hunk 1 above)
Comment (grote) [reply app/src/main/java/org/onionshare/android/tor/TorManager.kt:74]: I prefer to set STARTING earlier, so as soon as we called `start()` our state is STARTING already.

Here, STATUS_ON, counts down the startLatch, so we can create the HS and only once that has been published, we consider ourselves STARTED, so there's some divergence between the TorService states and the states we use for our own state management.

File: app/src/main/java/org/onionshare/android/ShareManager.kt (hunk 2 above)
Comment (grote) [reply app/src/main/java/org/onionshare/android/ShareManager.kt]: That shouldn't happen anymore after merging the PR that disables the button and after the extra commits that refactor and improve the state handling. Now there's a dedicated STOPPING state with a disabled button which should also prevent this.

File: app/src/main/java/org/onionshare/android/ShareManager.kt [hunk 3]
```diff
@@ -1,221 +1,155 @@
...
         } else {
-            _shareState.value = ShareUiState.FilesAdded(newList)
-        }
-    }
-
-    fun removeAll() {
-        // release persistable Uri permissions again
-        shareState.value.files.forEach { file ->
-            file.releaseUriPermission()
+            LOG.error("Unhandled state: ↑")
```
Comment (grote) [inline app/src/main/java/org/onionshare/android/ShareManager.kt:102]: This entire `shareState` method is quite a beast and does not handle all permutations of states. If there's any suggestions how to reconcile the states of the various sub-systems in a better way, please let me know!

File: app/src/main/java/org/onionshare/android/ShareManager.kt [hunk 4]
```diff
@@ -1,221 +1,155 @@
...
-                    ensureActive()
-                    _shareState.value = ShareUiState.Starting(files, state.progress, 0)
-                    if (state.complete) sendPage = getSendPage(state)
-                }
-                val page = sendPage ?: error("SendPage was null")
+        startSharingJob = coroutineScope {
+            launch(Dispatchers.IO) {
+                // call ensureActive() before any heavy work to ensure we don't continue when cancelled
                 ensureActive()
+                // When the current scope gets cancelled, the async routine gets cancelled as well
```
Comment (akwizgran) [inline app/src/main/java/org/onionshare/android/ShareManager.kt:131]: Does the same apply to torManager.start() below? What are the implications of cancelling that call?

**Comment by akwizgran** [comment]
I rebased this branch on master and left it running over the weekend. Tor started successfully, which hasn't happened on master in many attempts. Shortly after the onion address appeared I was able to open the download page in Tor Browser. I didn't download the file.

After about 24 hours OnionShare was still running and the onion address was still shown in the UI, but the hidden service was no longer reachable in Tor Browser. The same was true this morning (about 60 hours after starting OnionShare). When I tapped the stop button in OnionShare it got stuck in the "stopping" state, with the notification still showing "Currently sharing files". It stayed in this state for several hours. Exiting the app via the back button and relaunching didn't make a difference. Eventually I had to force-stop the app.

Unfortunately since the app was running for so long, all the useful information had expired from the system log. I'll try to reproduce this without leaving the app running for so long.

File: app/src/main/java/org/onionshare/android/ShareManager.kt [hunk 5]
```diff
@@ -1,221 +1,155 @@
...
 
-    @OptIn(DelicateCoroutinesApi::class)
     @Suppress("BlockingMethodInNonBlockingContext")
     private suspend fun startSharing() {
         if (startSharingJob?.isActive == true) {
             // TODO check if this always works as expected
             startSharingJob?.cancelAndJoin()
         }
+        shouldStop.value = false
         // Attention: We'll launch sharing in Global scope, so it survives ViewModel death
```
[the remark this answers is not in the document]
Comment (grote) [reply app/src/main/java/org/onionshare/android/ShareManager.kt:126]: Ahh good catch. Through some corners this actually inherits the ViewModelScope, so this needs to go back to GlobalScope.

File: app/src/main/java/org/onionshare/android/ShareManager.kt (hunk 4 above)
Comment (grote) [reply app/src/main/java/org/onionshare/android/ShareManager.kt:131]: Yes, there a `JobCancellationException` gets thrown which we catch and call `stop()` which stops Tor again.

**Pull request merged**